### ISSUE TITLE
[1.10] Porting

### ISSUE THREAD

**Issue by lehjr** [MachineMuse/MachineMusePowersuits#725]
I've started porting to 1.9.4. I've Got Numina mostly up and running . I still have to update the sound handler since sounds are now event based, and I need to see how TileEntities are synced since getDescriptionPacket() is no longer a thing. I'd like to bring back the customizations and the models from 1.7.10, but I'm not sure if it's even possible with the rendering/model changes. 

All that said, 1.9.4 is already a disappointment for me so far. This is with Numina up and running with no other mods installed, leaving nothing but Vanilla rendering and worldgen:
![2016-06-08_15 03 57](https://cloud.githubusercontent.com/assets/1084367/15916117/4bd2ed94-2dbd-11e6-9484-70a5e9ce2456.png)

Edit:
TODO List/progress:
Numina porting:
- [ ] Figure out the new sound system once MPS gets that far
- [ ] Networking tweaks 
- [ ] Rendering? (figure out what is and isn't possible)
- [ ] JEI recipe handler tweaks

MPS porting:
- [ ] New Power Fist models (one with normal pose and one with open hand for firing plasma cannon) - partially done in OBJ format, needs coloring
- [ ] Armor Rendering + customization (see Numina rendering above)
- [ ] Lux Capacitor model rendering and colorization (in 1.8.9 I had issues with the OBJ models and had to make my own in JSON format. Even then, I could not get the texture to stay aligned and had to use a single color texture for the thrown entity part).
- [x] Power Tool tips - These were missing in 1.8.9. I really need to bring this back in 1.9.4 since without them, it's nearly impossible to figure out how much charge each armor piece has. 
- [ ] Power Systems Compatibility - depending on what the mod landscape looks like, I might implement other forms of power. I don't even know if there's an official RF API yet.
- [ ] EV Capacitor - There's been a texture for it for some time now. As far as I can tell, it was an idea that was never implemented, probably because the armor has such a massive amount of power storage at its top tier already. It might be an idea though to use this as the top tier storage and rescale the lower tiers.
- [x] Running in Dev environment - in 1.7.10 building the mod from source was rather straight forward. Getting it to run in a dev environment was a nightmare. Going forward, it will be easier to use repos (maven/ivy/whatever) and update as needed. 
- [ ] Addons and external mod support - Supporting external mods through modules made running in a dev environment extremely difficult and heavy at times. For instance, the Chisel module supporting 2 conflicting versions of the mod required disabling one to get MPS to run in a dev environment (this was actually in MPS and not the addons). The Forestry plugin couldn't be updated because the newer version disabled support for some bee addon for Forestry. The EC2 and AE modules were RV1, then RV2/RV3 but had some issues when EC2 finally dropped the old AE API. The OmniWrench suddenly caused crashing after what appears to be CoFH-Lib being in CoFH-Core. Railcraft caused issues with running in a dev environment because they often bundle outdated API's of CoFH, IC2, and Buildcraft. On and on.. In summary, the Addons may see a return out of necessity, depending on the lifespan of 1.9.4.

**Comment by lehjr** [comment]
It's a vanilla bug: 
https://bugs.mojang.com/browse/MC-90602

**Comment by Lordmau5** [comment]
getUpdateTag() is for the initial packet and onDataPacket() is for when you receive new packets. 
Keep in mind that you need to write the NBT in getUpdatePacket() 

https://github.com/Lordmau5/FFS/blob/1.9/src/main/java/com/lordmau5/ffs/tile/abstracts/AbstractTankTile.java#L98

**Comment by lehjr** [comment]
@Lordmau5 thanks! That just leaves sound, which I'll figure out later. I'm hoping I can bring back the models from 1.7.10, but the rendering system is quite alien to me.

**Comment by Lordmau5** [comment]
https://mcforge.readthedocs.io/en/latest/effects/sounds/

I was looking for this for the past 30 minutes on Google just to find out that there was something on the Wiki about it, haha.

**Comment by lehjr** [comment]
Thanks again. Looks like I'll put that off until I start working on MPS itself.

**Comment by lehjr** [comment]
Looks like I'll need another pair of models for the power fist thanks to being able to switch the main hand. On a side note, I just love it when Scala tells me that a method overrides "some base member" but won't tell me where. I know the syntax isn't always easy to read, but that's just pointless.

**Comment by Lordmau5** [comment]
I just noticed this part:

_[...] the Chisel module supporting 2 conflicting versions of the mod [...]_

Just play it safe and go for support of [Chisel-Team's Chisel](https://github.com/Chisel-Team/Chisel/tree/1.9/dev). :)

//EDIT: Or just don't implement any addon-stuff at all and make a custom "base-addons" mod, which includes Chisel, Forestry, ... support :P
That way MPS can stay clean and other people can include the mod if they need / want it.

**Comment by Xen1290** [comment]
@lehjr For chisel @Lordmau5 is right, Chisel-Team Chisel is the right version to support.

For the dev environment: i think that Numina need to be incorporated into MPS like mps addons. In the past was necessary this division because MachineMuse developed Anima, but now isn't the case. Results: less problems for the dependencies, only one project, easy to run in a dev environment.

**Comment by lehjr** [comment]
The chisel module itself isn't even a consideration atm as I'm still working on getting the base of MPS itself ported. It wasn't even enabled in 1.8.9. Having Numina as a separate mod is not really an issue as it's just one dependency. The issue is having modules that support mods that either update and change their API's frequently or are slow to update between Minecraft releases (or don't update at all) such as Railcraft, BluePower, Applied Energistics/Extra Cells, Mekanism, Project Red, Thermal Expansion, MFR, probably others as well. Mod interaction is a great experience for the end user, but for the developer, it can be hell to deal with, especially when MC versions have shorter life spans. In order to move faster between versions, you have to travel lighter. As such, some of the modules may either be dropped or moved out into an addon package.

**Comment by Xen1290** [comment]
@lehjr I think that now addons are the last thing to do. As you said, MC versions have shorter life and not all mods update to newest versions, so you don't need to update those addons. Once the main mod is done, you can focuss to addons. For now you can disable those addons.

**Comment by Xen1290** [comment]
For Numina, i tried to setup a dev environment, and it was a pain, I've never got a working dev environment, instead for example immersive engineering was very easy, and this mod has a lot of dependencies to deal with, like MPS.
Keep in mind that i don't know java very well. Maybe it's only me.

**Comment by lehjr** [comment]
Any addons will depend on the mod ecosystem and the life span of the current version of MC. I have a feeling a few popular mods might have spent too much time milking 1.7.10 for all it's worth and adding even more things that will make porting even harder instead of focusing on the future. 

Integrating Numina back into MPS wouldn't make it any easier to run in a dev environment. It would just make more of a mess by increasing the size and complexity of MPS itself.

**Comment by lehjr** [comment]
If you're referring to getting the 1.7.10 dev environment setup for MPS, that's a different animal. The dependencies were only met for building, not running. On top of that, there were some issues like Railcraft using outdated API's and one of the SmartMoving libs had an incorrect reference that made it crash Minecraft. Then there were the 2 conflicting Chisel mods as dependencies, so you had to comment out one in the code. All of that is just another reason why I want to cut down the dependencies for MPS and go with addons where possible.

**Comment by Xen1290** [comment]
Yeah I'm referring to 1.7.10 and for trying to running a dev evironment.
Anyway for the rest, I'll see what you will do. 
I will follow your work.
Good Luck!

**Comment by lehjr** [comment]
And... Forge builds for MC1.10 have already started showing up.

**Comment by Xen1290** [comment]
Yeah, and MCMultipart will be included.
So.. Directly 1.10?

**Comment by lehjr** [comment]
I'll probably finish a working build on 1.9.4 first mainly because almost everything I have left is rendering stuff and the sound changes. I also have to create 2 more models for the power fist because of the ability to change the main hand and because of dual wielding. I still haven't textured the first 2. https://github.com/lehjr/PowersuitsModels/tree/master/PowerFist

**Comment by lirannl** [comment]
I heard that 1.10 is a very, very easy transition and that the Forge API barely changed at all.

**Comment by lehjr** [comment]
I'll probably start working on the sound stuff this week. After that I'll do the other 2 models I need followed by the rendering. Once I can get a working build done I'll look at 1.10. It just makes more sense for me to focus on getting a working build to port rather than porting untested code.

**Comment by lehjr** [comment]
OK, so on to 1.10 porting. No more MC version changes until I get everything ported over to the new rendering system.

**Comment by lehjr** [comment]
Numina is on 1.10 now, not quite finished as I still need to finish up the sound changes and clean up some old 1.7.10 code, but it's usable now.

**Comment by lehjr** [comment]
Working on weeding out the depreciated stuff, then it's time to experiment with Forge's working example of using Obj models with IModelPart to toggle visibility of model parts.

**Comment by lehjr** [comment]
Currently working on applying the original powerfist texture to the new OBJ models. I should be able to apply the texture to one model and reuse the mapping for the other 4.

**Comment by lehjr** [comment]
Finally got the power fist models rendering. I ported TheGreyGhost's  Item Transform Helper to 1.10.2 to help with the job. I still have to work on the textures, since model textures all have to be square now. (Edit: yes, this is the left hand equipped model since main hand can now be right or left hand. Right hand models is the same but for the right hand)
![2016-08-05_09 08 03](https://cloud.githubusercontent.com/assets/1084367/17440352/368ab90c-5afa-11e6-9bc7-254c706660c3.png)

**Comment by Xen1290** [comment]
Wow! Very good work lehjr

**Comment by lehjr** [comment]
In case anyone is wondering what the holdup is, it's the rendering, specifically the color customization. Due to changes in how models are handled and rendered, it's a bit more challenging than I would like. In 1.7.10, models were rendered "raw" and it was easy to change the color of a model or a part at render time. However, the "baked" model system is a static model system that doesn't play well with changing things after the models are already "baked". That said, I'm still pressing forward and trying to get things as close to what they were in 1.7.10. It's just taking far more time than it really should.

**Comment by azurewyvern** [comment]
All I can say is.....amazing.  I can't wait for you to finish as this mod was one I have been missing for a very long time.

**Comment by lehjr** [comment]
Lux Capacitor finally done, colors and rotation finally working. I scaled down the original models and created a new texture. The frame is being rendered with the builtin setup while the lens is being rendered with a TESR. Lens color change is done through material color change. Each uniquely colored lens model is cached for each model rotation. Now to get the armor to render.

**Comment by lehjr** [comment]
I really hate to do it, but it looks l like I'm going to have to port some of the scala code back to Java. It's given me nothing but issues and it's costing me time, not to mention it's damned hard to read at times.

**Comment by lehjr** [comment]
I guess while I'm at it, I'll see if I can optimize the network bandwidth a bit by using smaller data types where possible. For instance, instead of using 3 doubles (64 bits each) for each color, use 1 int (32 bits) for all 3 (8 bits each plus 8 bits for alpha). 

Edit: This change was made for Lux Capacitors, the armor is much more tricky, but much of that was already optimized.

**Comment by lehjr** [comment]
Looks like I haven't complained about the rendering system in a few days, so let me start off with an epic rant on that < insert epic rant here>. Moving on...

**Comment by eyeonus** [comment]
Considering all the awful that is the 1.8+ rendering system, have you looked in to render APIs?

This one https://github.com/jaquadro/Chameleon is apparently more like the 1.7.10 render method.

**Comment by lehjr** [comment]
Thanks. No, was actually trying to avoid relying on external API's as much as possible just because I have to wait on those API's in order to port to the next version. This is why I'm ditching SmartMoving. 

However, given how much time I have spent on rendering code and still working on it, I will check this API out. Anything that simplifies things is welcome.

**Comment by eyeonus** [comment]
All I know is, it's by the guy that does the Storage Drawers mod for 1.10, it's a required library for that.

According to him, "Chameleon Mod 1.10.2,1.10, 1.9.4,1.9, 1.8.9, 1.8 is a shared code library for Minecraft 1.8 and later.  Although it does not currently provide rendering hooks to restore the 1.7 render pipeline, it does provide a more comprehensive RenderBlocks replacement that is also suited for TESR rendering."

I linked to the github. If it's [open source|permission granted], you could internalize the code you need, assuming you wanted to put the effort in, or could show me what is you need to do. (I'm much better with code than graphics, but I don't know scala at all.)

**Comment by lehjr** [comment]
Well, I'm not good with Scala either and it's one of the things that has slowed down porting. I've actually toyed with the idea of porting everything to Java for that reason. However, the mod is much larger and more complex than it looks, and there are some areas where Scala is actually much simpler and easier to read, especially when dealing with traits and sealed traits.  

The big obstacle with rendering is that the old method used "raw" models and you could interchange them with ModelBase models. Models now have to have everything applied before baking, meaning that parts have to be colored by changing material colors, and parts have to be toggled on or off and then the model is "baked" and the rendering is done by a completely different setup that uses quads. Then you have to figure out a new key-value system to store these things because every change you make is a new model. Every change someone else makes is a new model too. Then you have to sync all this stuff between players.

**Comment by eyeonus** [comment]
Does the texture have to be baked in as well? It is possible to re-bake the models?

**Comment by lehjr** [comment]
When using the built in loader and baking system, textures are cached
for you, but not if you are calling the load/bake system yourself. In that
case you have to cache the textures before baking. I don't have the code in
front of me ATM, but textures are cached then a reference to the texture is
baked in, I think. The texture must be cached using the texture stitch
event. I think I ran into some issues with this because of the model bake
event firing first. You can however bake models without the event. Rebaking
the models causes the model in the model cache to change, so this is pretty
useless.

Currently I'm working on porting The 1.7.10 version of Numina to Java. I'm
gonna see if I can port both MPS and Numina 1.7.10 versions over this week
and go from there. This way it will be easier for people to contribute to
the project.

**Comment by eyeonus** [comment]
I ask because, if the renderer gets the texture externally, you could change the texture without needing to re-bake the model. If, on the other hand, the texture is part of the baked model, I got no ideas.

**Comment by lehjr** [comment]
texture is mapped as part of the model when it's baked. You can in some instances change the texture, but for MPS, textures don't change, just the part colors and the part visibility. But for every one of those changes, it requires a new baked model.

**Comment by eyeonus** [comment]
If it's possible to change the texture without re-baking the model, then you can do the color changing without needing to re-bake. A red texture is different than a blue texture, just apply the color values to the texture directly.

As far as part visibility is concerned, the only thing I can think of is to pre-bake every possible combination of pieces with a default white texture. Depending on how many parts, that could be a lot of models, but that's not nearly so bad as having to make a new model for every combination of parts and every possible color for every part.

**Comment by lehjr** [comment]
The colors aren't simple colors, they're 4 bytes meaning millions of
textures would be needed for that and they would have to be done as a
translucent overlay because the models already have textures.

**Comment by lehjr** [comment]
basically there are no simple solutions. I'm hoping to come up with some ideas while porting the mode fully over to Java.

**Comment by eyeonus** [comment]
Okay, I'm sorry, I must not be explaining things correctly. Use ONE texture, do the texture stitch caching thing, apply the color changes directly to THAT texture. All you need is the memory location of the cached texture, the original texture, and the ability to write to memory.

White texture -> Baked model texture cache
white texture (sample pixel = [0xff,0xff,0xff])-> color changes -> blue texture [sample pixel = [0x00,0x00,0xff])
blue texture -> write to cached memory location

Baked model texture cache = blue texture.

**Comment by lehjr** [comment]
If you change the color of the cached texture, it would be changed globally. Models use a single texture for all of the parts. Every part is able to be colored independently. By changing the texture color, you change them for every player model that your player sees.

**Comment by eyeonus** [comment]
What I'm suggesting would work for all of that except possible SMP-related issues with other people also having the armor, and I'm betting I could solve that too, but I'd need to look at the code. It would be easier if each piece had its own texture, but still possible without that, just more annoying code-wise.

Point me out to the code that does the texture making. I obviously don't have enough information on exactly how texturing works.

**Comment by lehjr** [comment]
All the textures are stitched into a single global sheet, including Minecraft textures. These textures are then accessed through a TextureAtlasSprite which contains coordinates and a name. What you're looking for is IRetexturableModel.  I don't have anything working on that at the moment because everything is in Scala, which is why I'm porting it over to Java in the first place. However, thinking back, IRetexturableModel I believe is an IModel, and if that's the case, then it does have to be baked anyway. 

In 1.7.10 the code is completely different. The colors are applied through openGL calls and each part is rendered individually. In 1.10.2, models are rendered as baked quads, and you normally use built in methods which are either an itemRenderer or a blockRenderer because even rendering is threaded now.

**Comment by eyeonus** [comment]
Okay. SMP would be difficult, it could be done for up to 'x' sets of armor,
but I don't have any idea for unlimited sets.

One texture per piece per set, direct memory access to the sheet, replace
the texture for the specific piece on the fly by modifying the source
texture and overwriting that spot on the sheet with the modification,
leaving the source texture unchanged.

Multiple pieces could use the same spot on the sheet if they're too small
to need a full one to themselves. Any parts that were a single color could
be represented by a single pixel in the source texture, for example.

**Comment by lehjr** [comment]
You can't change the sheet after the stitch event, and the stitch event is only fired once. It's not impossible to recolor models, but it's not easy. I did finish the Lux Capacitor, but I haven't uploaded the code because I started working on the armor, figuring it wouldn't be that horrible and the code blew up from there. 

Basically for the LuxCapacitor I'm using a hashmap and storing by color and model rotation. That's easy because there are only 2 parts and only one gets colored. The armor models have multiple parts, and multiple models. There are actually 2 sets of armor models with parts mapped to different parts of the player, at least in 1.7.10. In 1.10.2, nothing for that is working yet because all of it needs to change. At the moment, I'm still trying to think of a key-value setup for the map to store the baked models. I came up with a couple ideas but nothing concrete yet.

**Comment by lehjr** [comment]
I've made some pretty good progress porting the Scala code to Java. I'm doing this in 1.7.10 rather than 1.10.2 because it's easier to try something that already works. The biggest obstacles will be Scala Traits, especially those that extend other traits and interfaces and have working code because Java interfaces can do neither. Other issues include just deciphering some of the weird syntax and figuring out equivalent data types.

This is an unexpected delay, but unavoidable. Scala is great for those that are great with it. But it limits the  amount of help available because not everyone is proficient with Scala. It also does not decompile well at all in the IDE but gives vague references to code at best. On top of that, Scala errors do not show up until Java errors are fixed. So during porting, you may have 100 or so java errors to deal with, then find 500 more Scala errors afterwards. Scala does provide ways of doing things that Java still cannot, but as many as when the original code was written. There are also many Java to Scala converters available, but none going the other way around. I'll be switching to Java 1.7 support for Minecraft 1.7.10, since no one should be using Java 6 anymore. After that, it's Java 8.

**Comment by lehjr** [comment]
Porting from Scala to Java is taking longer than expected, however it has given me a better idea of what the code is actually doing. And the more of it I port to Java, the easier it is to read. I'm about 2/3 the way through at the moment. I did most of it by hand but in the past few days I have fallen on using a decompiler to help speed things along.

**Comment by lehjr** [comment]
Progress of 1.10.2 porting is still on hold until I finish porting the 1.17.10 version to Java. That said, I have been working steadily on that, and currently I'm working on porting the rendering code to Java. After that, is the rest of the Tinker Table GUI, followed by the network code.

**Comment by eyeonus** [comment]
I know this a bit of an old topic, but I had an idea regarding the problem with textures and the new rendering piepline:

1: Only have the original, more Minecraft-looking armor set.
2: For that armor set, make all the parts of the individual pieces that are coloured the same by default be the same piece- i.e. you can change _all_ the white parts on the chestplate blue, but you can't change the white part of the left arm to be blue and the white part of the right arm to be red, for example. This is to reduce the number of parts so that there aren't a seriously large number of combinations.
3: Only allow the 16 standard colours or invisible. Again, this is to reduce the number of combinations to something a bit more manageable.

I'm pretty sure that, even the people who would storm your castle with torches and pitchforks if there's no visual armor customization, would be placated with this. As for me, I usually run around with all but the neck piece (of the "modern" armor) or the spine (of the "classic" armor) invisible, so I don't really care too much about color.

**Comment by lehjr** [comment]
I haven't touched the 1.10.2 stuff in about a month now while I've been
porting the mod + Numina to Java. However I do now have a better handle on
the rendering and I should be back working on 1.10.2 by the end of the
month. Porting the mod has made things much easier to read, and given me
some ideas, and I should be done porting it either today or tomorrow,
followed by a few days of cleanup and testing, then a maintenance release
for 1.7.10. The 1.10.2 port will be done completely in Java.

**Comment by eyeonus** [comment]
Wow, you're further along than I thought you were. I figured the java porting would take longer. Good on you.

I'm not worried about it, I know you're not sitting on your thumbs on this, and I appreciate your hard work and dedication. I'm perfectly capable of being patient.

**Comment by lehjr** [comment]
MPS and Numina are both ported to pure Java.

**Comment by lehjr** [comment]
MPS and Numina are both ported to pure Java.

**Comment by Xen1290** [comment]
Great work @lehjr !

**Comment by lehjr** [comment]
OK, if anyone is still paying attention... in the 1.10.2/1.11 release, I'm looking at removing the weird keybinding GUI and just using the vanilla one. Anyone have any feedback on that?

**Comment by LeeLR** [comment]
The old GUI would always reset my bindings on me so I say go for it.  Thank you for all your hard work, this is one of my favorite mods.

**Comment by lehjr** [comment]
From a gameplay standpoitnt, I found the whole GUI setup weird to use. For me, it nice to just be able to look in the list with the rest of the keybinds. From a programming standpoint, it's actually better to use the vanilla setup too.

**Comment by Lordmau5** [comment]
I'd recommend going with vanilla keybind stuff.
Some recent MC version (iirc) allowed modifiers, so you can bind things like Ctrl+G or Alt+M to keys.
Either that, or it's always been in and I just never figured it out haha

Then again - you need to take into account that you have a couple modules that need to be configured then. Not that the cluttering would make the config GUI worse, considering that if you have a modpack with a bunch of mods, all of those add a couple themselves ;D

Though maybe allowing modules to be bound to the same key?
E.g. that I could disable the thrusters + jetpack with the same key?

**Comment by eyeonus** [comment]
Vanilla would probably be better, but instead of assigning keys to modules, how about one toggle key that gets held down to show a radial menu for toggling, and use the gui to group modules into the radial, in the interest of less bindings?

Kind of like the idea for the revamped fist module chooser on one of the other issue threads.

**Comment by eyeonus** [comment]
That said, as long as I can toggle multiple things on and off at the same time, and I don't need to reset the boss every time I launch the game, I don't care all that much.

**Comment by eyeonus** [comment]
Arg. Can't edit posts from phone.

*binds

**Comment by Lordmau5** [comment]
I mean, the old system could be kept, as long as that bug with "keybinds being lost upon game restart" is getting fixed.
I quite liked the GUI for it in the tile.

**Comment by lehjr** [comment]
The main reason for wanting to eliminate it is because of code complexity, not to mention it's the only place in the code where access transformers have to be used. Plus going to the Tinker table to figure out your keybindings never really seemed appealing. The options menu is portable.

**Comment by eyeonus** [comment]
Um, you don't have to go to the tinker table. By default the K key opens the key bind menu.

**Comment by eyeonus** [comment]
I don't even know HOW to see the key binds from the table.

And even if you were right, there's a module for the fist that makes the table only needed for installing and removing modules.

**Comment by lehjr** [comment]
The portable Tinker table module is not a full tinker table.

**Comment by eyeonus** [comment]
Yes, I know. You can't install or remove modules using it. You can do everything else the table does with it, however.

**Comment by lehjr** [comment]
The 1.7.10 pure java port is out as an alpha due to the amount of code changes. I'm about halfway through porting Numina to 1.10.2 now.

**Comment by Razorskills** [comment]
This is amazing! I will definitely do some testing!

--------
edit:

Heres a crash report after launching with:
 - CodeChickenLib-1.7.10-1.1.3.138-universal
 - CodeChickenCore-1.7.10-1.0.7.47-universal
 - ModularPowersuits-0.11.1.100
 - NotEnoughItems-1.7.10-1.0.5.120-universal
 - Numina-0.4.1.100

cheating in the armor, equipping them all using right click, then pressing e to open to the inventory. and boom: [Crash Report](https://www.dropbox.com/s/2k0tq01dgmpt2x9/crash-2016-11-20_20.38.10-client.txt?dl=0)

I will try doing each armor piece separately to see if there is a certain one that is the main cause. 

--------
I tested each piece separately, all of them are the same. Also, if you equip by right clicking, then press f5 to change to 3rd person, it will also close to: [Crash](https://www.dropbox.com/s/yonqgvpaszarewv/crash-2016-11-20_20.58.47-client.txt?dl=0)

I also tested with Smart Moving and same results with a few extra "Smart Moving" lines in the crash.

**Comment by lehjr** [comment]
Thanks. Looks like an issue with SmartMoving not being installed. I'll have to fix this before I get back to work on 1.10.2.

**Comment by lehjr** [comment]
@Razorskills The issue was the hard reference to the SmartMoving (RenderPlayerAPI)  model. Next build should be up shortly.

**Comment by Razorskills** [comment]
I just tested it without smart moving and it seems to be all good! (model wise)

I also gave it a test with smart moving, and as long as you also have Render Player API, the model hooks correctly. So that is good.

-----

Only other weird thing, which was always a bug before. when you go to change the visuals of the suit, the pieces become invisible, I don't quite remember if its fixable in the gui other than changing the model type.

But for some reason, I thought you had changed it back to the good old normal square texture, and not the crazy performance hitting 3D one. Oh well. Perhaps one day?

Anyways, I will play around with it and see if I can break anything later!

**Comment by lehjr** [comment]
I noticed that when porting. When you go to the visuals, the armor piece resets. to all invisible settings. You have to manually reset each section of the armor. I'm not entirely sure why that is.

The vanilla style armor is in 1.8.9, mainly because I couldn't even begin to figure out how the rendering was supposed to work, let alone make it work in 1.8.9 with the rendering changes in Minecraft. That's why I ported the mod to Java, so I could actually read and understand what the code was doing.

I do however want to add the vanilla style armor as an option in 1.10 and maybe even backport it if needed. The biggest issue I see is that it would need a different packet system, which means that in order to see a performance benefit instead of a hit, only one type of armor would be able to be allowed. Otherwise both types of packets would be sent around. The "3d" armor packet setup is quite complex, as every toggleable (spelling?) section of every armor piece and ever color setting is tracked so that all players can see the customized armor as the player wearing it does (third person view).

**Comment by lehjr** [comment]
I think I will change the recipe extraction and loading setup for 1.10. Instead of having the recipes extracted based on what mods are loaded, it would be easier to just extract them and load them based on what mods are loaded. Then there won't be people asking where to download them from.

**Comment by lehjr** [comment]
This is why I want to get rid of the weird GUI keybinding setup, so these kind of hackish workarounds aren't needed: https://github.com/MachineMuse/MachineMusePowersuits/blob/1.10.2-Java/src/main/java/net/machinemuse/powersuits/control/KeyBindingHelper.java

**Comment by lehjr** [comment]
wow, load times for 1.10.2 are pretty horrible.

**Comment by lehjr** [comment]
![2016-11-29_19 32 03](https://cloud.githubusercontent.com/assets/1084367/20734939/a661c2e0-b66a-11e6-926d-b367d7c617b7.png)

A little progress

**Comment by lehjr** [comment]
Only a few modules left to get working. I brought back the treetap module for IC2 rubber trees. The leaf blower will have it's settings merged into a single slider because there's really nothing logical about having different settings for leaves, plants, and snow.

**Comment by lehjr** [comment]
Power Fist rendering should be done this weekend. I'll be using the ModelBase model (modified for left and right hand) for this round. I'll look at Wavefront models for 1.11.

**Comment by lehjr** [comment]
![2016-12-24_23 57 43](https://cloud.githubusercontent.com/assets/1084367/21469791/5d481c60-ca35-11e6-914e-f3b56b186dc5.png)

**Comment by Lordmau5** [comment]
Holy damn, good job :O

**Comment by lehjr** [comment]
Lux capacitor rendering stuff is taking way longer than it should but is mostly done (still an issue with models missing on reload or join multiplayer). Armor rendering hopefully won't be this painful.

**Comment by eyeonus** [comment]
Keep it up man, you're doing great!

I volunteer for trib^K^K^K^Kalpha if needed.

**Comment by lehjr** [comment]
![2017-02-09_18 16 01](https://cloud.githubusercontent.com/assets/1084367/22807240/00e7a2a4-eef4-11e6-8096-9d9812df6170.png)

**Comment by lehjr** [comment]
Armor rendering stuff is next. That in itself will be a project.

**Comment by EzerArch** [comment]
May the coffee be with you. :D

**Comment by lehjr** [comment]
gonna take a lot of coffee and probably a few Irish coffees.

**Comment by Bluexin** [comment]
Wow great to see this mod still exists, it's always been one of my favs ^-^
Do you plan to add support for the now increasingly more used Tesla API? It'd help a lot :D (we're working on a server without IC2 nor TE for example, centered around Mekanism and enderIO -> no way to charge I guess?).
If you need help with that, don't hesitate to ask as I've worked a bit with it before (like fixing Mekanism's Tesla support).

**Comment by Lordmau5** [comment]
There's Forge's own energy API that could be used, _Forge Energy_.

I mean, adding support for Tesla won't be a big issue I'd assume - it's 1:1 to RF and FE, so yea :P

**Comment by lehjr** [comment]
Power really is not a concern just yet. Once rendering stuff is finished then I'll look at power as well as the availability, stability, and reliability of other mods in regards to our modules that extend or use features from other mods, like Railcraft and Apllied Energistics. The rendering stuff for the armor is going to take time due largely to the complexity of the customizable models, especially with things like glowing parts. However, once complete it hopefully won't have to get overhauled again in the foreseeable future.

**Comment by ToadsworthLP** [comment]
The Thermal* mods are coming back soon as KingLemming's almost finished with most stuff, so out-of-the-box RF supoort would be nice, too :)

**Comment by lehjr** [comment]
Out of the box working mod is the priority at the moment. Power stuff will be addressed after other things are fixed and working.

**Comment by lehjr** [comment]
Looks like I'll be working on the armor models before I can get them to even bake, let alone render. Fun. -_-

**Comment by lehjr** [comment]
First small victory. I set up a test block that I can use to render the armor components as a block to troubleshoot the models themselves. Boots model finally works (renders as a textured block. Armor rendering still waiting on fixing models so they actually load) (note: in case anyone is wondering, I'm not changing the UV layout of the armor models so if you have any custom textures from 1.7.10 they will still work): 
![2017-02-19_12 51 16](https://cloud.githubusercontent.com/assets/1084367/23104821/fc5ee92a-f6a2-11e6-944c-7a547adbacb7.png)

**Comment by EzerArch** [comment]
![image](https://68.media.tumblr.com/7a5b07b04c09a1b5d1ec229039cc678b/tumblr_n9o0wpMInf1sgl0ajo1_500.gif)

**Comment by lehjr** [comment]
armor leg models now rendering as a block. Once I get all of them actually working as a block I can finally start working on the armor rendering itself.

**Comment by lehjr** [comment]
Armor models are all fixed. Now to get them to render. Now that the models are actually working, this should go much easier.

**Comment by eyeonus** [comment]
\o/

**Comment by lehjr** [comment]
Armor almost kinda renders, well, it renders like something from a Cyriak video. Hopefully I can fix that by the end of the week.
Just in case you didn't get the reference: https://www.youtube.com/user/cyriak

**Comment by lehjr** [comment]
This is the issue I'm having: https://vimeo.com/205581567

**Comment by GamerFreak** [comment]
That sucks, I'm really glad to see this mod being maintained though! Good luck fixing the rendering.

**Comment by lehjr** [comment]
yeah, the main 2 things left to fix are the rendering and the recipe handler.

**Comment by GamerFreak** [comment]
Couldn't you go back to the basic armor rendering? Like in the 1.4.7 version of the mod? I'm no coder just a fan of the mod so excuse me if I come off as uneducated on the subject.

**Comment by lehjr** [comment]
I _could_ and it would be almost effortless, but it's one of those things I was trying to avoid. I did it in the 1.8.9 port but I wasn't fond of it. For me, the custom models are a big part of what makes the mod popular, the ability to customize the armor, both functionally and cosmetically.

**Comment by GamerFreak** [comment]
I understand, I do enjoy the cosmetic side of the mod quite a bit. :) Sorry I cannot be any help to figuring out the armor rendering!

**Comment by lehjr** [comment]
I'm confident that I'll figure it out. It's just going to take more time than I had hoped.

**Comment by GamerFreak** [comment]
Maybe a temporary or beta release of the mod with the older armor model would be a good idea? After you finish the recipe handler that is... Then you could work on the full release with all of it's functions.

**Comment by lehjr** [comment]
Not really. Once you have one system setup, you have to provide an upgrade path if you are going to change it. And I really don't want to end up supporting 2 forks of the mod, one with skin type armor and one with 3d style armor. It would actually better to just finish the rendering first, and release an alpha and let people try it with different things so that those bugs can be fixed while working on the recipe handler.

**Comment by lehjr** [comment]
The broken recipe handler can be worked around by the end user as it has in the past with MineTweaker/CraftTweaker scripts if needed, but the broken rendering has no end user workaround.

**Comment by lehjr** [comment]
I originally had similar rendering issues with the lux capacitor. In fact
it was responsible for one of the many delays in getting to this point. The
new rendering system is quite fragile and it doesn't take much for things
to go sideways. So it's best to let Minecraft and Forge to handle as much
as possible. Unfortunately this is one of those special cases were we can't
just create an item and map a static model to it.

**Comment by lehjr** [comment]
Rendering now closer to finished : 
![2017-02-28_17 26 58](https://cloud.githubusercontent.com/assets/1084367/23437229/49cdcab6-fddb-11e6-9121-1d9371039399.png)

**Comment by lehjr** [comment]
Armor now rendering correctly using a temporary method where the models are bound to "dummy items" and loaded that way through JSON files. Anyway, most things are working. I'll try and sort out the recipe handler next. Meanwhile, here's a video where you can see some of how far along the mod is: https://vimeo.com/206453129

**Comment by lehjr** [comment]
Models are now centered and facing the player in the Tinker Table cosmetic tab

**Comment by lehjr** [comment]
recipe handler seems to be working now. With any luck the first alpha will be out on Friday.

**Comment by EzerArch** [comment]
> recipe handler seems to be working now. With any luck the first alpha will be out on Friday.

NO! We still have 1.7.10 fixes to do. >:3

**Comment by lehjr** [comment]
> NO! We still have 1.7.10 fixes to do. >:3

So you're saying you want to be in charge of those fixes? Thanks!!! That gives me more time to work on other versions :D

**Comment by eyeonus** [comment]
Ezer is such a nice person, volunteering to take some of the load off for you.

**Comment by lehjr** [comment]
On the bright side, some of those 1.7.10 bugs are more likely to get fixed with a 1.10.2 version out. 

Also, I'm thinking about using the "cosmetic modules" to change the armor skin on the player, so basically if you turn off all the parts of your armor model you could be using the skin type armors or none if you don't install one of the modules. 

After there's a working alpha it might be nice to compare the network bandwidth usage to that of 1.17.10. I do have an idea that could substantially reduce the amount of NBT data used for rendering which involves using presets instead of tracking every color and visibility of each individual part of every model. Even with hundreds of thousands of possible presets, it would still be much less traffic.

**Comment by lehjr** [comment]
Thoughts: https://vimeo.com/207395952

**Comment by eyeonus** [comment]
I think you should have the sebK armor as the default, and the high poly armor as the installable module. Maybe have the Citizen armor as a skinswitch tinker setting for the default, instead of another module that needs installed?

Also, how does it handle having both modules installed? It looks like it doesn't crash or anything, but that doesn't mean nothing funky happens.

**Comment by lehjr** [comment]
@aionys That's kinda what I was thinking too. Actually, having the high poly armor as installable module fits in with what I was talking about with servers, where the module could be disabled. With that disabled, as well as disabling the nbt data that goes with it (not yet implemented), the network traffic should be substantially less. Right now, there isn't any framerate difference with any of the armor choices, at least none that I can tell in Linux where my dev environment is. 

As far as what happens when 2 are installed in the same armor piece, the current code just checks for one, and uses it if installed, if not, it moves on to the next. Once everything is finalized, I'll try to set it up so only one can be selected at a time.

**Comment by lehjr** [comment]
Cosmetic modules round 2 https://vimeo.com/207650675

**Comment by Bluexin** [comment]
Nice to see the amazing progress you're making on this great mod :D 
Can't wait to add it to my pack heheh.

**Comment by lehjr** [comment]
Everything is more or less ported. Really no need to keep this one open. The alpha development continutes here: https://github.com/MachineMuse/MachineMusePowersuits/issues/756
